src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets

```diff
@@ -121,8 +121,93 @@ Copyright (c) .NET Foundation. All rights reserved.
+
+    <!-- Delete the orphaned files. -->
+    <Delete
+        Files="@(_OrphanPublishFileWrites)"
+        TreatErrorsAsWarnings="true">
+      <Output TaskParameter="DeletedFiles" ItemName="_OrphanFilesDeleted"/>
+    </Delete>
+    
+    <!-- Write new list of current files back to clean file. -->
+    <WriteLinesToFile
+        File="$(IntermediateOutputPath)$(PublishCleanFile)"
+        Lines="@(_CurrentPublishFileWrites)" 
+        Overwrite="true"/>  
+  </Target>
+
+  <!--
+    ============================================================
+                                        _GetCurrentAndPriorPublishFileWrites
+    Get the list of files written in the previous publish and the list of files to be written in this publish.
+    ============================================================
+    -->
+  <Target Name="_GetCurrentAndPriorPublishFileWrites" >
+    <PropertyGroup>
+      <NormalizedPublishDir>$([MSBuild]::NormalizeDirectory($(PublishDir)))</NormalizedPublishDir>
+    </PropertyGroup>
+    
+    <Hash ItemstoHash="$(NormalizedPublishDir)">
+      <Output TaskParameter="HashResult" PropertyName="NormalizedPublishDirHash" />
+    </Hash>
+    <PropertyGroup>
+      <PublishCleanFile Condition="'$(PublishCleanFile)'==''">.$(NormalizedPublishDirHash.Substring(0, 10)).txt</PublishCleanFile>
+    </PropertyGroup>
+    
+    <!-- Read in writes made by prior publish. -->
+    <ReadLinesFromFile File="$(IntermediateOutputPath)$(PublishCleanFile)">
+      <Output TaskParameter="Lines" ItemName="_UnfilteredPriorPublishFileWrites"/>
+    </ReadLinesFromFile>
+
+    <ConvertToAbsolutePath Paths="@(_UnfilteredPriorPublishFileWrites)">
+      <Output TaskParameter="AbsolutePaths" ItemName="_UnfilteredAbsolutePriorPublishFileWrites"/>
+    </ConvertToAbsolutePath>
+  
+    <!-- Find all files in the final output directory. -->
+    <FindUnderPath Path="$(NormalizedPublishDir)" Files="@(_UnfilteredAbsolutePriorPublishFileWrites)" UpdateToAbsolutePaths="true">
+      <Output TaskParameter="InPath" ItemName="_PriorPublishFileWritesInOuput"/>
+    </FindUnderPath>
+    
+    <!-- Remove duplicates from files produced in the previous publish. -->
+    <RemoveDuplicates Inputs="@(_PriorPublishFileWritesInOuput)" >
+      <Output TaskParameter="Filtered" ItemName="_PriorPublishFileWrites"/>
+    </RemoveDuplicates>
+
+    <ItemGroup>
+      <_CurrentPublishFileWritesUnfiltered Include="@(ResolvedFileToPublish->'$(NormalizedPublishDir)%(RelativePath)')" Condition="'$(PublishSingleFile)' != 'true'"/>
+      <_CurrentPublishFileWritesUnfiltered Include="@(ResolvedFileToPublish->'$(NormalizedPublishDir)%(RelativePath)')" 
+                                           Condition="'$(PublishSingleFile)' == 'true' And '%(ResolvedFileToPublish.ExcludeFromSingleFile)' == 'true'"/>
+      <_CurrentPublishFileWritesUnfiltered Include="$(NormalizedPublishDir)$(AssemblyName)$(_NativeExecutableExtension)" Condition="'$(PublishSingleFile)' == 'true'"/>
```

Comment:
Is this dependent on `PublishSingleFile` or `UseAppHost`? I think my mental model might be confused.

Reply:
I was basing this off of the logic in [_GeneratePublishDependencyFile](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L876-L941), which uses `PublishSingleFile`. Is there somewhere else I should look for uses of `UseAppHost`?

Reply:
@rainersigwald it doesn't look to me like `UseAppHost` will work here- can you expand?

Reply:
Sorry, totally missed your question! This is my mental model, but it could easily be wrong:

* `UseAppHost` controls whether a native application named `$(AssemblyName)$(_NativeExecutableExtension)` (`projectname.exe`) is created in the output folder
* `PublishSingleFile` controls whether that application has all the applications dependencies bundled into it (`true`) or whether they're just next to it on disk (`false`). It requires `UseAppHost=true` (or you get `error NETSDK1067: Self-contained applications are required to use the application host. Either set SelfContained to false or set UseAppHost to true.`).

If that's right, then the question "was `foo.exe` produced in the output directory" would be better answered with `$(UseAppHost)`. But again: I'm not confident of that.

Reply:
You're right- just updated it. Thanks!